src/Compilers/Core/Portable/MetadataReader/MetadataTypeName.cs

```diff
@@ -30,16 +32,30 @@ internal partial struct MetadataTypeName
         /// </summary>
         private string _namespaceName;
 
+        /// <summary>
+        /// <see cref="ReadOnlyMemory{T}"/> version of <see cref="_namespaceName"/>.  Preferred when possible to avoid
+        /// the copy of the portion of <see cref="_fullName"/> used for <see cref="_namespaceName"/>.
+        /// </summary>
+        private ReadOnlyMemory<char> _namespaceNameMemory;
```

Comment:
>_namespaceNameMemory [](http://example.com/codeflow?start=37&length=20)

Would it make sense to introduce `readonly struct LazyString { string String => ...; ReadOnlyMemory<char> Memory; }` ?

Reply:
i would be ok with that.  will see what COmpiler thinks here.  Or if they'd prefer to keep the separation :)

Reply:
note: the compiler has slightly different semantics here as well for the strings, which might be hard to ensure with a LazyString abstraction.  Specifically, they ensure that some of the strings will be reference-equals in certain cases (for example, the TypeName and UnmangledTypeName in the event of non-generic type).  You don't get this if you have the LazyString wrapper, as each GetString() will produce a fresh string.  

You could potentially allow write access into the LazyString to force those values.  But it's tehn getting firmly into wonkytown IMO.  But still open :)

Reply:
I'd had the same thought about `LazyString` too. My thought was though that the current approach is holding onto memory it doesn't need to. The basic pattern is the following: 

```csharp
private ReadOnlyMemory<char> _nameMemory;
private string? _name;
public string Name => _name ??= _nameMemory.ToString();
```

Once the `Name` property is evaluated there is no more reason to hang onto `_nameMemory`. Doing so is potentially rooting a larger string in this type. Seems more efficient to make `_nameMemory = _name.AsMemory()`. That way it's self contained. At the point you're changing both items seems better to wrap into a type.

Reply:
so, i basically didn't go down this route because there are several use cases here and a few of the ones don't fit into that pattern.

First, NamespaceSegments (which is the big win) doesn't benefit from that approach here.  If NamespaceSegments are an `ImmutableArray<LazyString>`, then you cannot mutate those values within the array, so anyone trying to get them as strings, will allocate over and over again.

Second, we have the pattern of comparing and reusing the actual String instance from another LazyString in some cases.   So having a LazyString, with an externally mutable `string` value just felt like i was getting on hte bus to wonky town.  At that point, esp with the inability to use this for NamespaceSegments, i just felt like it was better to have the separate fields here.